### PR TITLE
Enable windows build

### PR DESCRIPTION
**Pull request by xantares**

### CODE REVIEW

**Comment by xantares** [comment]
you can start from here @edisongustavo

**Comment by xantares** [comment]
@edisongustavo I applied your changes here, and I updated the bld.bat to build bindings too.

**Comment by edisongustavo** [comment]
Nice.

File: recipe/bld.bat [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+:: language bindings are in a separate repo without releases
+git clone https://github.com/CGAL/cgal-swig-bindings.git
+cd cgal-swig-bindings
+
+:: this test requires numpy and we do not want to build-depend on it
+del examples/python/test_aabb2.py
+
+cmake -LAH -G"NMake Makefiles"                ^
+  -DCMAKE_PREFIX_PATH="%LIBRARY_PREFIX%"      ^
+  -DCMAKE_INSTALL_PREFIX="%PREFIX%"           ^
```
Comment (edisongustavo) [inline recipe/bld.bat]: Install in `%LIBRARY_PREFIX%` here too.

Comment (xantares) [reply recipe/bld.bat]: in fact this needs https://github.com/CGAL/cgal-swig-bindings/pull/79 for everything to be installed correctly as there's a dll and some python bindings. you're right for the dll, but not for the bindings. my PR will allow to set CMAKE_INSTALL_PREFIX="%LIBRARY_PREFIX%" while overriding python install dir.

**Comment by xantares** [comment]
@edisongustavo could you test my branch ?

**Comment by edisongustavo** [comment]
> @edisongustavo could you test my branch ?

Sure.

File: recipe/bld.bat (hunk 1 above)
Comment (edisongustavo) [reply recipe/bld.bat]: Yes, in fact you're correct.

**Comment by edisongustavo** [comment]
# CGAL only

Compiling the branch I got these errors:

```
W:\Miniconda\conda-bld\recipe_1481627996245\work\CGAL-4.9\include\CGAL/GMP/Gmpzf_type.h(403): error C2664: 'double __gmpz_get_d_2exp(mpir_si *,mpz_srcptr)': cannot convert argument 1 from 'CGAL::Gmpzf::Exponent *' to 'mpir_si *'
W:\Miniconda\conda-bld\recipe_1481627996245\work\CGAL-4.9\include\CGAL/GMP/Gmpzf_type.h(403): note: Types pointed to are unrelated; conversion requires reinterpret_cast, C-style cast or function-style cast
W:\Miniconda\conda-bld\recipe_1481627996245\work\CGAL-4.9\include\CGAL/GMP/Gmpzf_type.h(414): error C2664: 'double __gmpz_get_d_2exp(mpir_si *,mpz_srcptr)': cannot convert argument 1 from 'CGAL::Gmpzf::Exponent *' to 'mpir_si *'
W:\Miniconda\conda-bld\recipe_1481627996245\work\CGAL-4.9\include\CGAL/GMP/Gmpzf_type.h(414): note: Types pointed to are unrelated; conversion requires reinterpret_cast, C-style cast or function-style cast
W:\Miniconda\conda-bld\recipe_1481627996245\work\CGAL-4.9\include\CGAL/GMP/Gmpzf_type.h(425): error C2664: 'double __gmpz_get_d_2exp(mpir_si *,mpz_srcptr)': cannot convert argument 1 from 'CGAL::Gmpzf::Exponent *' to 'mpir_si *'
W:\Miniconda\conda-bld\recipe_1481627996245\work\CGAL-4.9\include\CGAL/GMP/Gmpzf_type.h(425): note: Types pointed to are unrelated; conversion requires reinterpret_cast, C-style cast or function-style cast
```

This is due to the `typedef mp_exp_t Exponent`.

So I believe I got confused with my first patch. The `typedef` should read like:

`typedef mp_si Exponent`

# cgal-swig-bindings

## bld.bat

After fixing that locally I tried to recompile. The build then stopped with:

```
(W:\Miniconda\conda-bld\recipe_1481628871708\_b_env) W:\Miniconda\conda-bld\recipe_1481628871708\work\CGAL-4.9\build\cgal-swig-bindings>cmake --build . --config "Release" --target INSTALL
Error: could not load cache
```

So to fix it I added a `cd ..` just after the cgal compilation, before the bindings.
My guess is that on windows, cmake does not support having inner directories with other cmake files. I believe I saw something like that in the past.

I also created a `build` directory inside the `cgal-swig-bindings`:

```
mkdir build
cd build
```

## compilation

First I got this linking error:

```
[ 32%] Linking CXX shared module ..\..\build-python\CGAL\_CGAL_Kernel.pyd
LINK : fatal error LNK1104: cannot open file 'python27_d.lib'
```

Whis is bizarre, why would it want to link against pyhon in debug?

I tried to solve it by passing `CMAKE_BUILD_TYPE=%CMAKE_CONFIG%` at the cmake command line but I wasn't successful.

I also noticed that it said "Installing configuration Debug", so I think that `nmake` is getting confused somehow.

As an alternative, I used the `ninja` package, which works quite well. We (at my company) use it internally and it is great. I know that Google uses it to build chromium too.

However, I ran into these compilation problems:

### [CGAL_Advancing_front_surface_reconstructionPYTHON_wrap.cxx](http://pastebin.com/ccC78Eq6) 

```
W:\Miniconda\conda-bld\recipe_1481630561811\work\CGAL-4.9\cgal-swig-bindings\build\build-python\CGAL\CGAL_Advancing_front_surface_reconstructionPYTHON_wrap.cxx : fatal error C1083: Cannot open compiler generated file: '': Invalid argument
```

I found this [link](https://community.rti.com/kb/why-do-i-get-fatal-error-c1083-cannot-open-compiler-generated-file-when-using-visual-studio) regarding this error and tried compiling with `ninja -j1`. It took forever and didn't solve the problem.

### [CGAL_Box_intersection_dPYTHON_wrap.cxx](http://pastebin.com/q8wayzQB)

```
build-python\CGAL\CGAL_Box_intersection_dPYTHON_wrap.cxx(3755): error C2976: 'std::vector': too few template arguments
```

Here you can see (at line 3755) that there is an usage of `std::vector` before the proper `#include`.
If I manually edit the file then the error disappears.

**Comment by xantares** [comment]
do you have logs ? for example the cmake output of the first issue

**Comment by edisongustavo** [comment]
> do you have logs ? for example the cmake output of the first issue

You mean the "Error: could not load cache" right?

I removed your `cd ..` commit and this is what happens: https://gist.github.com/edisongustavo/f6c90ca65007771b5ef29c0c3d454216

File: recipe/bld.bat [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+if errorlevel 1 exit 1
+
+cd ..
+
+:: language bindings are in a separate repo without releases
+git clone https://github.com/CGAL/cgal-swig-bindings.git
+cd cgal-swig-bindings
+
+:: this test requires numpy and we do not want to build-depend on it
+del examples/python/test_aabb2.py
```
Comment (edisongustavo) [inline recipe/bld.bat]: windows `del` command does not understand the `/` syntax. Change it to use `\`

File: recipe/bld.bat [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+  ..
+if errorlevel 1 exit 1
+cmake --build . --config %CMAKE_CONFIG% --target INSTALL
+if errorlevel 1 exit 1
+
+cd ..
+
+:: language bindings are in a separate repo without releases
+git clone https://github.com/CGAL/cgal-swig-bindings.git
+cd cgal-swig-bindings
```
Comment (edisongustavo) [inline recipe/bld.bat]: You should add here:

```
mkdir build
cd build
```

**Comment by xantares** [comment]
thanks, I see it builds in debug mode, I added CMAKE_BUILD_TYPE=Release.

File: recipe/bld.bat (hunk 3 above)
Comment (xantares) [reply recipe/bld.bat]: ok done

**Comment by edisongustavo** [comment]
Hi, this is the current log of the build after your last commits (`cd build`): https://gist.github.com/edisongustavo/082a24f2c2b273968be773eda3983977

**Comment by xantares** [comment]
ok see https://github.com/CGAL/cgal-swig-bindings/pull/80

**Comment by xantares** [comment]
@edisongustavo could you retry with my latest commit ?

**Comment by edisongustavo** [comment]
Hi, I tried with the latest commit and still got the `W:\Miniconda\conda-bld\recipe_1481630561811\work\CGAL-4.9\cgal-swig-bindings\build\build-python\CGAL\CGAL_Advancing_front_surface_reconstructionPYTHON_wrap.cxx : fatal error C1083: Cannot open compiler generated file: '': Invalid argument` error.

See: https://gist.github.com/edisongustavo/690e1874d563f7a7f3952c9216b38f08

**Comment by xantares** [comment]
could you try to build bindings in-source (without a build directory) ?

**Comment by edisongustavo** [comment]
> could you try to build bindings in-source (without a build directory) ?

Done just that and the same error happens again:

```
W:\Miniconda\conda-bld\._1481651417353\work\CGAL-4.9\cgal-swig-bindings\build-python\CGAL\CGAL_Advancing_front_surface_reconstructionPYTHON_wrap.cxx : fatal error C1083: Cannot open compiler generated file: '': Invalid argument
NMAKE : fatal error U1077: 'C:\PROGRA~2\MICROS~3.0\VC\bin\amd64\cl.exe' : return code '0x1'
Stop.
NMAKE : fatal error U1077: '"C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\BIN\amd64\nmake.exe"' : return code '0x2'
Stop.
NMAKE : fatal error U1077: '"C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\BIN\amd64\nmake.exe"' : return code '0x2'
Stop.
```

**Comment by edisongustavo** [comment]
Using the Visual Studio generator I believe I found the problem:

```
"W:\Miniconda\conda-bld\._1481652512338\work\CGAL-4.9\cgal-swig-bindings\INSTALL.vcxproj" (default target) (1) ->
"W:\Miniconda\conda-bld\._1481652512338\work\CGAL-4.9\cgal-swig-bindings\ALL_BUILD.vcxproj" (default target) (3) ->
"W:\Miniconda\conda-bld\._1481652512338\work\CGAL-4.9\cgal-swig-bindings\SWIG_CGAL\Advancing_front_surface_reconstruction\_CGAL_Advancing_front_surface_reconstruction.vcxproj" (default target) (11) ->
(InitializeBuildStatus target) ->
  C:\Program Files (x86)\MSBuild\Microsoft.Cpp\v4.0\v140\Microsoft.CppBuild.targets(312,5): error MSB3491: Could not write lines to file "_CGAL_Advancing_front_surface_reconstruction.dir\Release\_CGAL_Ad.6E718FE4.tlog\_CGAL_Advancing_fr
ont_surface_reconstruction.lastbuildstate". O caminho especificado, o nome do arquivo ou ambos são muito longos. O nome de arquivo totalmente qualificado deve ter menos de 260 caracteres e o nome do diretório menos de 248 caracteres. [W
:\Miniconda\conda-bld\._1481652512338\work\CGAL-4.9\cgal-swig-bindings\SWIG_CGAL\Advancing_front_surface_reconstruction\_CGAL_Advancing_front_surface_reconstruction.vcxproj]
```

The Windows is in portuguese, sorry about that.

The translation for:

>  "_O caminho especificado, o nome do arquivo ou ambos são muito longos. O nome de arquivo totalmente qualificado deve ter menos de 260 caracteres e o nome do diretório menos de 248 caracteres._"

is

> **The specified path, the filename or both are too long. The fully qualified filename must contain less than 260 characters and the name of the directory less than 248 characters**".

**Comment by xantares** [comment]
nice catch. maybe clone in the parent dir and with a shorter clone dir name: ie "git clone url dir" ?

**Comment by edisongustavo** [comment]
So I did a small "hack" so that this path limitation can be avoided. See this [gist](https://gist.github.com/edisongustavo/3d6bbc98b29db5f281aa08c3e3f0ec50)

With this I can compile everything fine.

I will post the results of the tests (still compiling) as soon as possible

**Comment by xantares** [comment]
what's that some kind of symlink ?

**Comment by edisongustavo** [comment]
> what's that some kind of symlink ?

Close...It's a "virtual drive". It's one of the strategies that I use to avoid this path limitation when compiling libraries. We use that all the time in our CIs.

**Comment by xantares** [comment]
maybe we could also try to clone the bindings in SRC_DIR and with a shorter name

**Comment by xantares** [comment]
and then, does the ctest command pass ?

**Comment by edisongustavo** [comment]
Not yet...there is something wrong with the compilation. I have to check.

```
W:\workspace\guga-etk-compile-sci20-py3\cgal-feedstock\recipe>W:\Miniconda\conda-bld\._1481656501206\_b_env\python.exe
Python 3.5.2 |Continuum Analytics, Inc.| (default, Jul  5 2016, 11:41:13) [MSC v.1900 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import CGAL
>>> from CGAL.CGAL_Kernel import Point_2
Traceback (most recent call last):
  File "W:\Miniconda\conda-bld\._1481656501206\_b_env\lib\site-packages\CGAL\CGAL_Kernel.py", line 18, in swig_import_helper
    return importlib.import_module(mname)
  File "W:\Miniconda\conda-bld\._1481656501206\_b_env\lib\importlib\__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 986, in _gcd_import
  File "<frozen importlib._bootstrap>", line 969, in _find_and_load
  File "<frozen importlib._bootstrap>", line 958, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 666, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 577, in module_from_spec
  File "<frozen importlib._bootstrap_external>", line 906, in create_module
  File "<frozen importlib._bootstrap>", line 222, in _call_with_frames_removed
ImportError: Module use of python27.dll conflicts with this version of Python.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "W:\Miniconda\conda-bld\._1481656501206\_b_env\lib\site-packages\CGAL\CGAL_Kernel.py", line 21, in <module>
    _CGAL_Kernel = swig_import_helper()
  File "W:\Miniconda\conda-bld\._1481656501206\_b_env\lib\site-packages\CGAL\CGAL_Kernel.py", line 20, in swig_import_helper
    return importlib.import_module('_CGAL_Kernel')
  File "W:\Miniconda\conda-bld\._1481656501206\_b_env\lib\importlib\__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
ImportError: No module named '_CGAL_Kernel'
```

**Comment by xantares** [comment]
yes, you have a python 3.5 interpreter, if the module is linked with python27.dll this is bad, at the cmake stage with one is detected ?

**Comment by edisongustavo** [comment]
Yes, you are correct. The problem happened because I have a Miniconda 2 installation, but I'm building a Python 3 build.

This shouldn't be a problem, however, there is a small problem in the `CMakeLists.txt` of cgal-swig-bindings: https://github.com/CGAL/cgal-swig-bindings/blob/master/CMakeLists.txt#L67

The calls should be switched, you should always call `find_package(PythonInterp)` before `find_package(PythonLibs)`, [docs of cmake saying this](https://cmake.org/cmake/help/v3.5/module/FindPythonLibs.html).

So I switched those lines and it linked correctly.

I've opened a [PR](https://github.com/CGAL/cgal-swig-bindings/pull/81)

**Comment by xantares** [comment]
ok, https://github.com/CGAL/cgal-swig-bindings/pull/82

**Comment by edisongustavo** [comment]
So I finally could compile it correctly.

This is the [bld.bat](https://gist.github.com/edisongustavo/3d6bbc98b29db5f281aa08c3e3f0ec50) file that I used.
This is the output of `ctest`:

```
Test project Z:/cgal-swig-bindings/build
      Start  1: pythontest_AABB_polyhedron_facet_intersection_example
 1/28 Test  #1: pythontest_AABB_polyhedron_facet_intersection_example ...   Passed    1.03 sec
      Start  2: pythontest_AABB_triangle_3_example
 2/28 Test  #2: pythontest_AABB_triangle_3_example ......................   Passed    0.07 sec
      Start  3: pythontest_Point_set_processing_3_test
 3/28 Test  #3: pythontest_Point_set_processing_3_test ..................   Passed    0.37 sec
      Start  4: pythontest_Polyhedron_incremental_builder
 4/28 Test  #4: pythontest_Polyhedron_incremental_builder ...............   Passed    0.06 sec
      Start  5: pythontest_advancing_front_reconstruction
 5/28 Test  #5: pythontest_advancing_front_reconstruction ...............   Passed    0.07 sec
      Start  6: pythontest_conforming
 6/28 Test  #6: pythontest_conforming ...................................   Passed    0.15 sec
      Start  7: pythontest_hds_incremental_builder
 7/28 Test  #7: pythontest_hds_incremental_builder ......................   Passed    0.07 sec
      Start  8: pythontest_hds_prog_halfedge_iterator
 8/28 Test  #8: pythontest_hds_prog_halfedge_iterator ...................   Passed    0.06 sec
      Start  9: pythontest_mesh_global
 9/28 Test  #9: pythontest_mesh_global ..................................   Passed    0.08 sec
      Start 10: pythontest_mesh_polyhedral_domain
10/28 Test #10: pythontest_mesh_polyhedral_domain .......................   Passed   16.19 sec
      Start 11: pythontest_polyhedron_modifier_example
11/28 Test #11: pythontest_polyhedron_modifier_example ..................   Passed    0.06 sec
      Start 12: pythontest_polyhedron_prog_cube
12/28 Test #12: pythontest_polyhedron_prog_cube .........................   Passed    0.06 sec
      Start 13: pythontest_regular_3
13/28 Test #13: pythontest_regular_3 ....................................   Passed    0.13 sec
      Start 14: pythontest_remove_outliers_example
14/28 Test #14: pythontest_remove_outliers_example ......................   Passed    0.08 sec
      Start 15: pythontest_simple_triangulation_3
15/28 Test #15: pythontest_simple_triangulation_3 .......................   Passed    0.07 sec
      Start 16: pythontest_test
16/28 Test #16: pythontest_test .........................................   Passed    0.07 sec
      Start 17: pythontest_test_aabb
17/28 Test #17: pythontest_test_aabb ....................................   Passed    0.07 sec
      Start 18: pythontest_test_as2
18/28 Test #18: pythontest_test_as2 .....................................   Passed    0.09 sec
      Start 19: pythontest_test_box_intersection_d
19/28 Test #19: pythontest_test_box_intersection_d ......................   Passed    0.07 sec
      Start 20: pythontest_test_convex_hull_2
20/28 Test #20: pythontest_test_convex_hull_2 ...........................   Passed    0.07 sec
      Start 21: pythontest_test_operators
21/28 Test #21: pythontest_test_operators ...............................   Passed    0.06 sec
      Start 22: pythontest_test_pmp
22/28 Test #22: pythontest_test_pmp .....................................   Passed    0.13 sec
      Start 23: pythontest_test_polyline_simplification_2
23/28 Test #23: pythontest_test_polyline_simplification_2 ...............***Failed    0.09 sec
TypeError: 'PS2_Constrained_Delaunay_triangulation_plus_2' object is not iterable

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "Z:/cgal-swig-bindings/examples/python/test_polyline_simplification_2.py", line 53, in <module>
    CGAL_Polyline_simplification_2.simplify(cdt, cid1, Squared_distance_cost(), Stop_below_count_ratio_threshold(0.5)) #simplify only
SystemError: <built-in function simplify> returned a result with an error set

      Start 24: pythontest_test_rt2
24/28 Test #24: pythontest_test_rt2 .....................................   Passed    0.25 sec
      Start 25: pythontest_test_sp
25/28 Test #25: pythontest_test_sp ......................................   Passed    0.08 sec
      Start 26: pythontest_test_t2
26/28 Test #26: pythontest_test_t2 ......................................   Passed    0.07 sec
      Start 27: pythontest_triangulation_prog1
27/28 Test #27: pythontest_triangulation_prog1 ..........................   Passed    0.07 sec
      Start 28: pythontest_vd_2_point_location
28/28 Test #28: pythontest_vd_2_point_location ..........................   Passed    0.09 sec
```

**Comment by xantares** [comment]
this is expected, this test also fails on the linux side, see https://github.com/CGAL/cgal-swig-bindings/issues/77

**Comment by xantares** [comment]
@conda-forge/core appveyor seems stuck, could you cancel the job queue ?

**Comment by ocefpaf** [comment]
I am looking into this now. It seems that only staged-recipes is building :unamused:

**Comment by xantares** [comment]
@edisongustavo appveyor builds are ok on my fork now: I used a shorter path rather than your hack, and I disabled py27 builds

**Comment by edisongustavo** [comment]
@xantares Great, I'm looking forward to it!

> I disabled py27 builds

Is python 2 not supported by `cgal-swig-bindings` ?

**Comment by xantares** [comment]
more precisely building with visual studio 9 is not supported,
I got one iterator type conversion error in one swig module.

**Comment by xantares** [comment]
see https://ci.appveyor.com/project/xantares/cgal-feedstock/build/1.0.1/job/x2iwji05dgmd7p25

**Pull request merged**